### PR TITLE
Added Tabby's modelling work

### PR DESCRIPTION
**Pull request by conor-dempsey-london**

### CODE REVIEW

**Review by conor-dempsey-london — COMMENTED** [review]
For the purposes of making reviews easier and code clearer I think it might be good for us to get into the habit of describing in the docstring the expected format of dataframes that are passed as function arguments. For example in your create labels function I'm not certain if the logic makes sense (it probably does) because I don't know, for example, what the 'month_year' column is exactly.

File: highstreets/features/processing_functions.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
+import pandas as pd
+from sklearn.linear_model import LinearRegression
+
+
+# create function to take averages of every x months
+# and see which hs fall in which proportion x months after that
+def create_labels(hs, label_month, comparison_start, comparison_end):
+    """Labels high streets at risk, depending on whether
+    spending is in which quartile of average of x months prior"""
+    hs_latest = hs.loc[(hs["month_year"] == label_month)]
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:10]: Is 'month_year' a specific month in a specific year, and is label_month also then supposed to specify month and year?

File: highstreets/features/processing_functions.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        "M"
+    )
+    hs_compare = hs.loc[(hs["month_year"]).isin(date_range)]
+    hs_compare = hs_compare.groupby(["highstreet_name"]).mean()
+
+    # merge dataframe
+    merged_df = pd.merge(hs_latest, hs_compare, on="highstreet_name")
+    # condtional labelling, if latest value not at least at average level, label at risk
+    merged_df["proportion"] = (
+        merged_df["txn_amt_wd_retail_x"] / merged_df["txn_amt_wd_retail_y"]
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:25]: I think this is another example where we should think about how to incorporate assumptions about the format of passed dataframes into our code more clearly. This function depends on particular columns existing that are only being referenced several lines into the function. It would be good to make these requirements of the passed dataframe explicit maybe.

File: highstreets/features/processing_functions.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    )
+    merged_df["labels"] = pd.qcut(
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:27]: Could this be simplified to pd.qcut(merged_df["proportion"], 4, labels=[0,1,2,3]) ?

Comment (conor-dempsey-london) [reply highstreets/features/processing_functions.py:27]: I'm also not completely clear how this achieves the functionality the function docstring describes. hs_latest will contain the means of all columns in the initial data frame averaged by highstreet for the label month, and hs_compare will contain means by highstreet during the comparison period for all columns. then the merged data frame takes a proportion of two particular columns and does a qcut on that proportion. Where does the latest value get compared to the comparison period explicitly?

File: highstreets/features/processing_functions.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+def create_mean_sd_o2(highstreet_df, predictor_days):
+    df = []
+    for days in predictor_days:
+        df.append(highstreet_df.loc[(highstreet_df["count_date"] == days)])
+    df = pd.concat(df)
+
+    # group to create mean and std across all months
+    mean = df.groupby(["highstreet_name"]).mean().reset_index()
+    # rename and merge so less possibility of mistakes
+    mean = mean.rename({"h13": "mean_o2"}, axis=1)
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:48]: Again maybe a good idea to just document assumptions about particular columns existing in the incoming data frame, h13 in this case, and highstreet_name (this isn't something I've done consistently, but probably a good habit for us to develop!)

File: highstreets/features/processing_functions.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        4,
+        labels=[0, 1, 2, 3],
+    )
+    merged_df = merged_df.reset_index()
+    # subset relevant columns
+    merged_df = merged_df[["labels"]]
+    return merged_df
+
+
+# loop and select data for relevant months
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:38]: Is this comment accurate? I'm not sure what it refers to

File: highstreets/features/processing_functions.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    mean = mean.rename({"h13": "mean_o2"}, axis=1)
+    std = df.groupby(["highstreet_name"]).std().reset_index()
+    std = std.rename({"h13": "std_o2"}, axis=1)
+    # merge and select columns
+    features = pd.merge(mean, std, on="highstreet_name")[
+        ["highstreet_name", "mean_o2", "std_o2"]
+    ]
+    return features
+
+    # loop and select data for relevant months
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:57]: Same q about this comment

File: highstreets/features/processing_functions.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    df = []
+    for months in predictor_months:
+        df.append(highstreet_df.loc[(highstreet_df["month_year"] == months)])
+    df = pd.concat(df)
+
+    # group to create mean and std across all months
+    mean = df.groupby(["highstreet_name"]).mean().reset_index()
+    # rename and merge so less possibility of mistakes
+    mean = mean.rename({"txn_amt_wd_retail": "mean"}, axis=1)
+    std = df.groupby(["highstreet_name"]).std().reset_index()
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:70]: Small thing but you could consider making a groupby object and then extracting from that object everything you need (mean and std here) instead of doing the groupby repeatedly.

File: highstreets/features/processing_functions.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+def create_gradient(hs, months):
+    """Use linear regression to calculate gradient of each high street
+    over the specified range of months"""
+    df = []
+    # create df of 3 months
+    for month in months:
+        df.append(hs.loc[(hs["month_year"] == month)])
+    df = pd.concat(df)
+    # craete ordinal time data
+    df["date_ordinal"] = pd.to_datetime(df["week_start"]).apply(
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:88]: I didn't know about toordinal, very handy!

File: highstreets/features/processing_functions.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        lambda date: date.toordinal()
+    )
+
+    # create list of df grouped by highstreet name, ordered
+    df_grad = []
+    names = []
+    for highstreet, subdf in df.groupby(["highstreet_name"], sort=True):
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:95]: This is a much cleaner way of extracting the gradients than what I had done! V good, uses pandas nicely. I was doing some awkward extracting of an array of all the data for each period I wanted gradients and looping over rows, this method, grouping by highstreet, is much better

File: highstreets/features/processing_functions.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        linear_regressor = LinearRegression()
+        # perform linear regression
+        model = linear_regressor.fit(X, y)
+        # make predictions
+        gradient = model.coef_
+        df_grad.append(gradient)
+        names.append(highstreet)
+
+    # flatten and create dataframe
+    flat_list = [item for sublist in df_grad for item in sublist]
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:108]: I feel like this way of making the data frame with highstreet names and gradients is a little bit opaque. Can you just make a data frame with the columns you want at the start and then append rows as you go?

File: highstreets/features/processing_functions.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    flat = [item for sublist in flat_list for item in sublist]
+
+    # create df
+    gradients = pd.DataFrame(
+        list(zip(names, flat)), columns=["highstreet_name", "gradient"]
+    )
+    return gradients
+
+
+def create_gradient_o2(hs, months):
```
Comment (conor-dempsey-london) [inline highstreets/features/processing_functions.py:118]: Maybe not worth it but this function is very similar to the previous one. I feel like if you specify some arguments for the relevant column names you might be able to make this one function. That would make it a bit more general and reusable and also force you to be explicit about which columns you expect the dataframe to have, or make column names arguments that you pass.

**Review by conor-dempsey-london — COMMENTED** [review]
This was good, code was generally clear and the logic seemed sound throughout. A few small comments that I think are worth considering. I think in particular the issue of writing general functions where possible and asking for expected columns to be passed as arguments, and/or having some checks to make sure certain required columns are present in a data frame, or in general making sure that data frames follow required formats, is a good idea.
